File: ClaudeRelayAndroid/core-protocol/src/main/kotlin/relay/protocol/SessionState.kt
package relay.protocol

/**
 * Lifecycle state of a ClaudeRelay session.
 *
 * Ports `SessionState.swift`. The [raw] values are the hyphenated wire strings
 * (e.g. `active-attached`); decoding is done via [fromRaw], which falls back to
 * [CREATED] for unknown values, matching the resilient client behaviour.
 */
enum class SessionState(val raw: String) {
    CREATED("created"),
    STARTING("starting"),
    ACTIVE_ATTACHED("active-attached"),
    ACTIVE_DETACHED("active-detached"),
    RESUMING("resuming"),
    EXITED("exited"),
    FAILED("failed"),
    TERMINATED("terminated"),
    EXPIRED("expired");

    /** Whether this state is terminal (no further transitions allowed). */
    val isTerminal: Boolean
        get() = when (this) {
            EXITED, FAILED, TERMINATED, EXPIRED -> true
            else -> false
        }

    /** Returns whether a transition from this state to [target] is valid. */
    fun canTransition(target: SessionState): Boolean = when (this) {
        CREATED -> target == STARTING
        STARTING -> target in startingTransitions
        ACTIVE_ATTACHED -> target in attachedTransitions
        ACTIVE_DETACHED -> target in detachedTransitions
        RESUMING -> target in resumingTransitions
        EXITED, FAILED, TERMINATED, EXPIRED -> false
    }

    companion object {
        private val startingTransitions = setOf(ACTIVE_ATTACHED, FAILED)
        private val attachedTransitions = setOf(ACTIVE_DETACHED, EXITED, FAILED, TERMINATED)
        private val detachedTransitions =
            setOf(RESUMING, ACTIVE_ATTACHED, EXPIRED, EXITED, FAILED, TERMINATED)
        private val resumingTransitions = setOf(ACTIVE_ATTACHED, FAILED, TERMINATED)

        fun fromRaw(value: String): SessionState = entries.firstOrNull { it.raw == value } ?: CREATED
    }
}

File: Sources/CodeRelayKit/Models/SessionState.swift
import Foundation

/// Represents the lifecycle state of a CodeRelay session.
public enum SessionState: String, Codable, Sendable {
    case created = "created"
    case starting = "starting"
    case activeAttached = "active-attached"
    case activeDetached = "active-detached"
    case resuming = "resuming"
    case exited = "exited"
    case failed = "failed"
    case terminated = "terminated"
    case expired = "expired"

    /// Whether this state is terminal (no further transitions allowed).
    public var isTerminal: Bool {
        switch self {
        case .exited, .failed, .terminated, .expired:
            return true
        default:
            return false
        }
    }

    // Static transition sets — allocated once, not per call.
    private static let startingTransitions: Set<SessionState> = [.activeAttached, .failed]
    private static let attachedTransitions: Set<SessionState> = [.activeDetached, .exited, .failed, .terminated]
    private static let detachedTransitions: Set<SessionState> = [.resuming, .activeAttached, .expired, .exited, .failed, .terminated]
    private static let resumingTransitions: Set<SessionState> = [.activeAttached, .failed, .terminated]

    /// Returns whether a transition from this state to the given target state is valid.
    public func canTransition(to target: SessionState) -> Bool {
        switch self {
        case .created:
            return target == .starting
        case .starting:
            return Self.startingTransitions.contains(target)
        case .activeAttached:
            return Self.attachedTransitions.contains(target)
        case .activeDetached:
            return Self.detachedTransitions.contains(target)
        case .resuming:
            return Self.resumingTransitions.contains(target)
        case .exited, .failed, .terminated, .expired:
            return false
        }
    }
}
